include/deal.II/matrix_free/matrix_free.h

```diff
@@ -802,6 +802,138 @@ class MatrixFree : public Subscriptor
+   * operations that depend on the result of the current cell loop in `dst` or
+   * want to modify `src`. The efficiency of caching depends on the numbering
+   * of the degrees of freedom because of the granularity of the ranges.
+   *
+   * @param cell_operation Pointer to member function of `CLASS` with the
+   * signature <tt>cell_operation (const MatrixFree<dim,Number> &, OutVector &,
+   * InVector &, std::pair<unsigned int,unsigned int> &)</tt> where the first
+   * argument passes the data of the calling class and the last argument
+   * defines the range of cells which should be worked on (typically more than
+   * one cell should be worked on in order to reduce overheads).
+   *
+   * @param owning_class The object which provides the `cell_operation`
+   * call. To be compatible with this interface, the class must allow to call
+   * `owning_class->cell_operation(...)`.
+   *
+   * @param dst Destination vector holding the result. If the vector is of
+   * type LinearAlgebra::distributed::Vector (or composite objects thereof
+   * such as LinearAlgebra::distributed::BlockVector), the loop calls
+   * LinearAlgebra::distributed::Vector::compress() at the end of the call
+   * internally.
+   *
+   * @param src Input vector. If the vector is of type
+   * LinearAlgebra::distributed::Vector (or composite objects thereof such as
+   * LinearAlgebra::distributed::BlockVector), the loop calls
+   * LinearAlgebra::distributed::Vector::update_ghost_values() at the start of
+   * the call internally to make sure all necessary data is locally
+   * available. Note, however, that the vector is reset to its original state
+   * at the end of the loop, i.e., if the vector was not ghosted upon entry of
+   * the loop, it will not be ghosted upon finishing the loop.
+   *
+   * @param operation_before_loop This functor can be used to perform an
+   * operation on entries of the `src` and `dst` vectors (or other vectors)
+   * before the operation on cells first touches a particular DoF according to
+   * the general description in the text above. This function is passed a
+   * range of the locally owned degrees of freedom on the selected
+   * `dof_handler_index_pre_post` (in MPI-local numbering).
+   *
+   * @param operation_after_loop This functor can be used to perform an
+   * operation on entries of the `src` and `dst` vectors (or other vectors)
+   * after the operation on cells last touches a particular DoF according to
+   * the general description in the text above. This function is passed a
+   * range of the locally owned degrees of freedom on the selected
+   * `dof_handler_index_pre_post` (in MPI-local numbering).
+   *
+   * @param dof_handler_index_pre_post Set which of the underlying DoFHandler
+   * objects passed at initialization should be selected by the
+   * `operator_before_loop` and `operation_after_loop`. Defaults to the
+   * `dof_handler_index` 0.
```

Comment:
I don't understand this parameter.

Reply:
MatrixFree can be initialized with a list of `DoFHandler` objects, e.g. one for velocity and one for pressure in a Stokes system. However, the call-back only gives me a range on one vector. The alternative to choosing which `DoFHandler` to provide a vector range for would be to return ranges to all `DoFHandlers` (by separate call-backs). In that case, the user would be given a third index, say 'dof_handler_index', specifying to which range that belongs. However, that feels like a bit of a waste because a user will not often run block systems through a single loop (otherwise, I would put the components in the same DoFHandler and vector). What do you think?